Refactor Index Management and support removal of Mixed Indexes (#3362)
JanusGraph/janusgraph: docs/schema/index-management/index-lifecycle.md

## Index states and transitions

The diagram below shows the possible states of an index in JanusGraph.

> Transitions which are not recommended or should be used with caution are shown in grey.

Reviewer: I don't like the formulation "are not recommended" very much to be honest. Maybe "are not recommended in production" or something like that?

![States and transitions](index-lifecycle.svg){: style="width:100%;"}

## States (SchemaStatus)
An index can be in one of the following states:

**INSTALLED**

:   The index has been created on the current JanusGraph instance, but it is not yet known to all instances in the cluster.
    This state is not shown in the diagram above because it is not stable in terms of synchronization between instances.

**REGISTERED**

:   The index is known to all instances in the cluster.
    However, it is not in use yet.

>     In this state, the index contains no data and is not eligible to be used in queries.

Reviewer: ```suggestion
    In this state, the index contains no data and will not be used in queries.
```

(My reasoning here is that JanusGraph is used by many people who aren't native English speakers so we should make the docs easy to understand.)

**ENABLED**

:   The index is enabled and can be used to answer queries.
    Furthermore, this is the only state in which insertions and deletions in the graph are forwarded to the index.

**DISABLED**

:   The index is disabled and will not be used to answer queries.
It also does no longer receive graph updates.
Therefore, it is recommended to use `REINDEX` to re-enable the index.

**DISCARDED**

:   The index has been decommissioned and its contents have been deleted.
    The only remaining artifact is the schema vertex representing the index.

## Actions (SchemaAction)
The following actions can be performed on an index to change its state via `mgmt.updateIndex()`:

**REGISTER_INDEX**

> :   Registers the index with all instances in the graph cluster.

Reviewer: I think graph cluster is a confusing term. The next sentence even speaks of _graph instances_ which should be _JanusGraph instances_.
Not sure what to do with _graph cluster_. Maybe just drop the _graph_?

After an index is installed, it must be registered with all graph instances.

**REINDEX**
:   Re-builds the index from the ground up using the data stored in the graph.
Depending on the size of the graph, this action can take a long time to complete.
Reindexing is possible in all stable states and automatically enables the index once finished.

**ENABLE_INDEX**
:   Enables the index so that it can be used by the query processing engine.
An index must be registered before it can be enabled.
If enabling the index manually instead of performing a reindex, be aware that past modifications of the graph are not represented in the index.
Enabling a previously disabled index without `REINDEX` may cause index queries to return ghost vertices, which have been deleted from the graph while the index was disabled.

**DISABLE_INDEX**

> :   Disables the index temporarily so that it is no longer used to answer queries.

Reviewer 2: Why is `DISABLE_INDEX` from `ENABLED` to `DISABLED` not in grey, but `ENABLE_INDEX` from `DISABLED` to `ENABLED` in grey?

Author: The idea behind the grey transitions was to give an intuition for consequences to be aware of. Disabling an index does exactly what everybody probably expects, but plain enabling without reindexing could lead to faulty index responses. One has to keep in mind that ENABLE does not just reverse a prior DISABLE and that may not be obvious to everybody at first glance.

The index also stops receiving updates in the graph.

**DISCARD_INDEX**
:   Removes all data from the index, preparing it for removal.
    This operation is supported for all composite indices and for some mixed index backends.
    For more information on index removal see [Removal](index-removal.md).

> **MARK_DISCARDED**

Reviewer 2: Two questions:
1. Why is this action marked grey?
2. What happens if the user doesn't delete the real indices from the index backend manually, but still runs this action?

Author: Let me answer 2. first:
If the user performs this step without deleting the indexed data, all of it remains in the index backend of course. At the same time, it looks to JanusGraph as there this index does not exist and may never have. Therefore, JanusGraph would happily allow creating a new index with the same name that already exists in the backend. Even though I have not yet bothered to find out exactly __how__ things misbehave, I still consider this highly unsafe.

Now to 1.:
Grey means that an operation has to be used with care and may carry unforeseen consequences. In the case of this action, the user must be aware of the implications I explained in the previous answer (2)

Reviewer 2: That makes sense. Would be even better if you could add a note/warning that one should not do `MARK_DISCARDED` without physically removing the indices.

:   All indices which can not be discarded using **DISCARD_INDEX** have to be discarded manually in the backend.
    As this is not recognized by JanusGraph automatically, the state can be set to **DISCARDED** manually via this action.

> **DROP_INDEX**

Reviewer: Wouldn't it be possible to merge the drop action into the two discard actions and then remove the `DISCARDED` state completely? Or what's the purpose of the `DISCARDED` state except of waiting for the user to execute this action?

Author: I already tried that, but there is currently no mechanism in place to properly wait for the long running index removal and schedule the deletion of the internal vertex (aka `DROP_INDEX`) afterwards.

Reviewer: Thanks for the explanation. Then I'm of course good with this. We can still try to improve this at some point in the future, but then it's out of scope for this PR.

:   Enables the index so that it can be used by the query processing engine.
An index must be registered before it can be enabled.
If enabling the index manually instead of performing a reindex, be aware that past modifications of the graph are not represented in the index.
Enabling a previously disabled index without `REINDEX` may cause index queries to return ghost vertices, which have been deleted from the graph while the index was disabled.

**DISABLE_INDEX**
:   Disables the index temporarily so that it is no longer used to answer queries.
The index also stops receiving updates in the graph.

**DISCARD_INDEX**
:   Removes all data from the index, preparing it for removal.
    This operation is supported for all composite indices and for some mixed index backends.
    For more information on index removal see [Removal](index-removal.md).

**MARK_DISCARDED**
:   All indices which do not support **DISCARD_INDEX** have to be discarded manually in the backend.
    As this is not recognized by JanusGraph automatically, the state can be set to **DISCARDED** manually via this action.
    This operation is supported for all composite indices and for some mixed index backends.

>     For more information on index removal see [Removal](index-removal.md).

Author: `MARK_DISCARDED` can be used with all index backends, including Elasticsearch. I agree that the comment is misleading and I will fix that. Thanks for the heads-up
